timemachine/cpp/src/nonbonded_interaction_group.cu

```diff
@@ -351,6 +341,82 @@ void NonbondedInteractionGroup<RealType>::execute_device(
     }
 }
 
+template <typename RealType>
+void NonbondedInteractionGroup<RealType>::verify_row_indices(const std::vector<int> &row_indices) {
+    if (row_indices.size() == 0) {
+        throw std::runtime_error("idxs can't be empty");
+    }
+    std::set<int> unique_idxs(row_indices.begin(), row_indices.end());
+    if (unique_idxs.size() != row_indices.size()) {
+        throw std::runtime_error("atom indices must be unique");
+    }
+    if (*std::max_element(row_indices.begin(), row_indices.end()) >= N_) {
+        throw std::runtime_error("indices values must be less than N");
+    }
+    if (*std::min_element(row_indices.begin(), row_indices.end()) < 0) {
+        throw std::runtime_error("indices values must be greater or equal to zero");
+    }
+}
+
+template <typename RealType>
+void NonbondedInteractionGroup<RealType>::set_atom_idxs(const std::vector<int> &atom_idxs) {
+    this->verify_row_indices(atom_idxs);
+    std::vector<unsigned int> unsigned_idxs = std::vector<unsigned int>(atom_idxs.begin(), atom_idxs.end());
+
+    std::set<unsigned int> unique_row_atom_idxs(unique_idxs(unsigned_idxs));
+    // compute set of column atoms as set difference
+    std::vector<unsigned int> col_atom_idxs_v = get_indices_difference(N_, unique_row_atom_idxs);
+    std::vector<unsigned int> row_atom_idxs_v(set_to_vector(unique_row_atom_idxs));
+    DeviceBuffer<unsigned int> d_col(col_atom_idxs_v.size());
+    DeviceBuffer<unsigned int> d_row(row_atom_idxs_v.size());
+    d_col.copy_from(&col_atom_idxs_v[0]);
+    d_row.copy_from(&row_atom_idxs_v[0]);
+
+    cudaStream_t stream = static_cast<cudaStream_t>(0);
+    this->set_atom_idxs_device(col_atom_idxs_v.size(), row_atom_idxs_v.size(), d_col.data, d_row.data, stream);
+    gpuErrchk(cudaStreamSynchronize(stream));
+}
+
+// set_idxs_device is for use when idxs exist on the GPU already and are used as the new idxs to compute the neighborlist on.
+template <typename RealType>
+void NonbondedInteractionGroup<RealType>::set_atom_idxs_device(
+    const int NC,
+    const int NR,
+    unsigned int *d_in_column_idxs,
+    unsigned int *d_in_row_idxs,
+    const cudaStream_t stream) {
+    if (NC + NR != N_) {
+        throw std::runtime_error("Total of indices must equal N");
+    }
+    const size_t tpb = warp_size;
+
+    // The indices must already be on the GPU and are copied into the potential's buffers.
+    gpuErrchk(cudaMemcpyAsync(
+        d_col_atom_idxs_, d_in_column_idxs, NC * sizeof(*d_col_atom_idxs_), cudaMemcpyDeviceToDevice, stream));
+    gpuErrchk(cudaMemcpyAsync(
+        d_row_atom_idxs_, d_in_row_idxs, NR * sizeof(*d_row_atom_idxs_), cudaMemcpyDeviceToDevice, stream));
+
+    // The neighborlist does not use the indices directly, rather it takes a continuous set of indices and the ixn group
```

Comment:
Let me know if this is unclear, as this feels unintuitive using the end of the column list for the row atom indices and vice versa.

Reply:
```suggestion
    // The neighborlist does not use the indices directly, rather it takes a contiguous set of indices and the ixn group
```
(I think it should be contiguous, right?)